**EXTRACT (COLLECTING THE DATA)**

**Importing Libraries**

In [ ]:
# Cell 1: Import the tools (libraries) we need
import requests      # This tool allows Python to 'talk' to the API website
import pandas as pd  # This tool allows us to organize data into tables
from datetime import datetime # This tool helps us get the current date and time

print("Tools imported successfully. Ready for Step 2!")

Tools imported successfully. Ready for Step 2!


**Setting the API Key**

In [ ]:
# To set API key here
API_KEY = "YOUR_API_KEY_HERE"

print("API Key saved!")

API Key saved!


**Defining Cities and Base URL**

In [ ]:
# Set the cities to track
# We have included Port Harcourt, Accra, and Cape Town
cities = ["Port Harcourt", "Accra", "Cape Town"]

# This is the web address to get the data
base_url = "http://api.openweathermap.org/data/2.5/weather"

# This is an empty list to store the data later
weather_data = []

print("Cities list (Port Harcourt, Accra, Cape Town) is ready!")

Cities list (Port Harcourt, Accra, Cape Town) is ready!


**The Collection Loop**

In [ ]:
# Create the collection loop
for city in cities:
    params = {"q": city, "appid": API_KEY, "units": "metric"}
    response = requests.get(base_url, params=params)

    if response.status_code == 200:
        data = response.json()
        weather_data.append({
            "City": data["name"],
            "Temperature": data["main"]["temp"],
            "Humidity": data["main"]["humidity"],
            "Condition": data["weather"][0]["description"],
            "Wind_Speed": data["wind"]["speed"],
            "Date_Time": datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        })
        print(f"Collected: {city}")

Collected: Port Harcourt
Collected: Accra
Collected: Cape Town


**TRANSFORM DATA (PROCESSING THE DATA)**

**Creating the Data Table**

In [ ]:
# Show the results
print(weather_data)

[{'City': 'Port Harcourt', 'Temperature': 27.76, 'Humidity': 73, 'Condition': 'moderate rain', 'Wind_Speed': 4.38, 'Date_Time': '2026-07-14 11:46:58'}, {'City': 'Accra', 'Temperature': 26.02, 'Humidity': 80, 'Condition': 'light rain', 'Wind_Speed': 6.08, 'Date_Time': '2026-07-14 11:46:58'}, {'City': 'Cape Town', 'Temperature': 16.73, 'Humidity': 77, 'Condition': 'few clouds', 'Wind_Speed': 0.89, 'Date_Time': '2026-07-14 11:46:58'}]


**Displaying the Data for Comparison**

In [ ]:
# Create the table
df = pd.DataFrame(weather_data)

# Show the clean table
print(df)

            City  Temperature  Humidity      Condition  Wind_Speed  \
0  Port Harcourt        27.76        73  moderate rain        4.38   
1          Accra        26.02        80     light rain        6.08   
2      Cape Town        16.73        77     few clouds        0.89   

             Date_Time  
0  2026-07-14 11:46:58  
1  2026-07-14 11:46:58  
2  2026-07-14 11:46:58  


In [ ]:
# Displaying the data for comparison
# Looking at relevant fields to compare the three cities
print(df[['City', 'Temperature', 'Humidity', 'Condition']])

            City  Temperature  Humidity      Condition
0  Port Harcourt        27.76        73  moderate rain
1          Accra        26.02        80     light rain
2      Cape Town        16.73        77     few clouds


**Identifying Insights**

In [ ]:
# Identifying the city with the highest humidity
# Comparing humidity values to determine which city is the most humid
max_humidity_row = df.loc[df['Humidity'].idxmax()]
print(f"City with highest humidity: {max_humidity_row['City']} at {max_humidity_row['Humidity']}%")

City with highest humidity: Accra at 80%


**LOAD (STORING THE DATA)**

In [ ]:
# Task 3: Load Data
# Save the table as a CSV file
df.to_csv("weather_data.csv", index=False)

print("Data saved successfully as 'weather_data.csv'!")

Data saved successfully as 'weather_data.csv'!


**KEY INSIGHTS**



*   **Temperature Variations**: My analysis shows a clear temperature gap between the cities. Port Harcourt and Accra are experiencing similar, warmer weather, with temperatures around 27.76°C and 26.02°C respectively. Cape Town is significantly cooler, recorded at 16.73°C, which highlights the geographic and climatic diversity between these locations.



*   **Humidity Observations**: After comparing the humidity levels, I identified that Accra has the highest humidity at 80%. Port Harcourt follows closely with 73%, while Cape Town sits at 77%. This indicates that while temperatures vary greatly, high humidity is a common factor across these selected African cities at this time.


*   **Weather Conditions**: There is a noticeable contrast in current weather patterns. Both Port Harcourt and Accra are experiencing precipitation ("moderate rain" and "light rain"), whereas Cape Town is seeing clearer skies with "few clouds." This demonstrates the value of real-time API data in tracking simultaneous, localized weather events.




**CONCLUSION**

This project successfully demonstrates the effectiveness of an ETL (Extract, Transform, Load) pipeline for real-time data analysis. By automating the extraction of weather data via the OpenWeather API, I was able to transform raw, unstructured responses into a clean, organized Pandas DataFrame, which allowed for efficient comparison and analysis. Storing this information as a CSV file provides a reliable foundation for future tracking or long-term trend analysis. Ultimately, this pipeline shows how organizations can move from raw data collection to actionable insights in a structured and reproducible way.